gammapy/utils/scripts.py

```diff
@@ -133,23 +140,70 @@ def _configure_root_logger(level='info', format=None):
 
 
 def read_yaml(filename, logger=None):
-    """Read config from YAML file."""
+    """
+    Read config from YAML file 
+    """
     import yaml
     if logger is not None:
         logger.info('Reading {}'.format(filename))
     with open(filename) as fh:
         config = yaml.safe_load(fh)
+
     return config
 
 
 def write_yaml(config, filename, logger=None):
-    """Write YAML config file
+    """
+    Write YAML config file
 
     This function can be used by scripts that alter the users config file.
+
+    Parameters
+    ----------
+    config : dict
+        config to write
+    filename : str, `~gammapy.exter.pathlib.Path`
+        file to write
     """
     import yaml
-    filename = filename + '.yaml'
+    filename = make_path(filename)
     if logger is not None:
         logger.info('Writing {}'.format(filename))
-    with open(filename, 'w') as outfile:
+    with open(str(filename), 'w') as outfile:
         outfile.write(yaml.dump(config, default_flow_style=False))
+
+
+def make_path(path):
+    """
+    Expand environment varibles on `~pathlib.Path` construction
+
+    Parameters
+    ----------
+    path : str, `~gammapy.extern.pathlib.Path`
+        path to expand
+    """
+    return Path(expandvars(str(path)))
+
+
+def recursive_update_dict(old, new):
```

Comment:
Update suggests in-place modification, like `dict.update`.
You call `copy`, i.e. the inputs are unmodified, right?

In this case I'd suggest calling the function `merge_dict` instead and also to explain in the docstring better that a new merged dict is returned and the inputs are unmodified.
(Example: https://github.com/ImmobilienScout24/yamlreader/blob/master/src/main/python/yamlreader/yamlreader.py#L24)

Actually: does this function even work properly? You call it recursively and it calls copy each time. I'm not saying it doesn't work, it's just not obvious to me.

Adding an `Example` section with one or two examples to the docstring and putting those as unit tests would be very helpful.

Although ... we need a more full-featured solution for config handling instead of building such utilities piece by piece.
I'll make an issue to discuss options shortly and then we can switch to that for 0.5. So @joleroi, please fix the bug at least for 0.4, but don't spend too much time on this.